pkg/fqdn/doc.go

```diff
@@ -12,20 +12,238 @@
+// DNS lookups in order to then allow L3 connections with the response
+// information. These must separately allowed with toFQDNs L3 rules. The
+// example above is a common "visibility" policy that allows all requests but
+// ensures that they traverse the proxy. This information is then placed in the
+// per-Endpoint and global DNS lookup caches and propagates from there.
+//
+// === L3 DNS ===
+// L3 DNS rules control L3 connections and not the DNS requests themselves.
+// They rely on DNS lookup cache information and it must come from the DNS
+// proxy, via a L7 DNS rule, or from the DNS poller (deprecated in cilium 1.8).
+//
+//  - toFQDNs:
+//      - matchName: "my-remote-service.com"
+//      - matchPattern: "bucket.*.my-remote-service.com"
+//
+// IPs seen in a DNS response (i.e. the request was allowed by a L7 policy)
+// that are also selected in a DNS L3 rule matchPattern or matchName have a /32
+// or /128 CIDR identity created. This occurs when they are first passed to the
+// toFQDN selectors from NameManager. These identities are not special in any
+// way and can overlap with toCIDR rules in policiies. They are placed in the
+// node-local ipcache and in the policy map of each endpoint that is allowed to
+// connect to them (i.e. defined in the L3 DNS rule).
+// Notes:
+// - Generally speaking, toFQDNs can only handle non-cluster IPs. In-cluster
+//   policy should use toEndpoints and toServices. This is partly historical but
+//   is because of ipcache limitations when mapping ip->identity. Endpoint
+//   identities can clobber the FQDN IP identity
+// - Despite being tracked per-Endpoint. DNS lookup IPs are collected into a
+//   global cache. This is historical and can be changed
+//   The original implementation created policy documents in the policy
+//   repository to represent the IPs being allowed and could not distinguish
+//   between endpoints. The current implementation uses selectors that also do
+//   not distinguish between Endpoints. There is some provision for this,
+//   however, and it just requires better plumbing in how we place data in the
+//   Endpoint's datapath
+//
+// === Caching, Long-Lived Connections & Garbage Collection ===
+// DNS requests are distinct traffic from the connections that pods make with
+// the response information. This makes it difficult to correlate one DNS
+// lookup to a later connection; a pod may reuse the IPs in a DNS response an
+// arbitrary time after the lookup occurred, even past the DNS TTL. The
+// solution is multi-layered for historical reasons:
+// - Keep a per-Endpoint cache that can be stored to disk and restored on
+//   startup. These caches apply TTL expiration and limit the IP count per domain.
+// - Keep a global cache to combine all this DNS information and send it to the
+//   policy system. This cache applies TTL but not per-domain limits.
+//   This causes a DNS lookup in one endpoint to leak to another!
```

Comment:
:thinking:

Reply:
There were many hand-wringey converstions about this. I think the conclusion was that:
1- Trusting the DNS data itself would use other tools (including controlling the DNS lookups and where they can go)
2- It's a little messy anyway because the DNS TTL is much shorter than a connection could be and we would expect to allow the whole connection. So letting it leak between endpoints is sort of like that (hand-wave).

Now that we have more wiring to the connection tracking we can clamp down on this quite a bit. Setting aside weird case like one pod doing service lookup and giving the IP to another (which I planned to support as a sibling to the DNS discovery) it's unlikely that leaking out of a pod will be needed.